### PR TITLE
formats.ini: Introduce `listToValue` argument

### PR DESCRIPTION
**Pull request by infinisil**
###### Motivation for this change

Allows coercing lists to values. E.g.

```nix
let
  pkgs = import ./. {};
  format = pkgs.formats.ini {
    listToValue = pkgs.lib.concatMapStringsSep ", " (pkgs.lib.generators.mkValueStringDefault {});
  };
in format.generate "config.ini" {
  foo.bar = [ 10 true ];
}
```

to create
```ini
[foo]
bar=10, true
```

For https://github.com/NixOS/nixpkgs/pull/120440#discussion_r619571093

Ping @aanderse @dotlambda @andir 

###### Things done

- [x] Tested manually
- [x] Added new test, running all format tests successfully with `nix-build pkgs/pkgs-lib/tests`
- [x] Updated docs

### CODE REVIEW

File: pkgs/pkgs-lib/formats.nix [hunk 1]
```diff
@@ -70,16 +70,23 @@ rec {
...
+            description = singleIniAtom.description + " or a list of them";
           }
         else
           singleIniAtom;
 
     in attrsOf (attrsOf iniAtom);
 
-    generate = name: value: pkgs.writeText name (lib.generators.toINI args value);
+    generate = name: value:
+      let
```
Comment (dotlambda) [inline pkgs/pkgs-lib/formats.nix:83]:
```suggestion
      assert !listsAsDuplicateKeys || listToValue == null;
      let
```

File: pkgs/pkgs-lib/formats.nix [hunk 2]
```diff
@@ -70,16 +70,23 @@ rec {
...
+        transformedValue =
+          if listsAsDuplicateKeys || listToValue == null then value
+          else lib.mapAttrs (section: lib.mapAttrs (key: val:
+            if lib.isList val then listToValue val else val
+          )) value;
```
Comment (dotlambda) [inline pkgs/pkgs-lib/formats.nix:88]:
```suggestion
          if listToValue != null then
            lib.mapAttrs (section: lib.mapAttrs (key: val:
              if lib.isList val then listToValue val else val
            )) value
          else
            value;
```

File: pkgs/pkgs-lib/formats.nix [hunk 3]
```diff
@@ -56,7 +56,7 @@ rec {
       };
     };
 
-  ini = { listsAsDuplicateKeys ? false, ... }@args: {
+  ini = { listsAsDuplicateKeys ? false, listToValue ? null, ... }@args: {
```
Comment (dotlambda) [inline pkgs/pkgs-lib/formats.nix:59]: add a comment documenting the type of `listToValue`

File: pkgs/pkgs-lib/formats.nix (hunk 1 above)
Comment (Infinisil) [reply pkgs/pkgs-lib/formats.nix:92]: Moved the assertion to the format definition start

**Review by dotlambda — APPROVED** [review]

**Review by andir — APPROVED** [review]

**Comment by Infinisil** [comment]
I noticed that there was a problem: Empty lists would be generated as well. E.g.

```nix
let
  pkgs = import ./. {};
  format = pkgs.formats.ini {
    listToValue = pkgs.lib.concatMapStringsSep ", " (pkgs.lib.generators.mkValueStringDefault {});
  };
in format.generate "config.ini" {
  foo.bar = [];
}
```

would generate
```ini
[foo]
bar=
```

I now added a commit to handle empty lists by removing the key completely. So this now returns this instead:

```ini
[foo]
```

I think this makes more sense. Does that sound good @andir @dotlambda?

**Comment by dotlambda** [comment]
Since it's not obvious what the semantics of empty lists should be, I'd rather we disallow them by using `types.nonEmptyListOf`. That might break existing configurations using `listsAsDuplicateKeys` though.
If we allow them, I actually prefer the former behaviour, i.e. an empty list corresponding to an empty string.

**Comment by Infinisil** [comment]
@dotlambda Sounds good, did that change

File: pkgs/pkgs-lib/formats.nix [hunk 4]
```diff
@@ -72,14 +81,27 @@ rec {
       iniAtom =
         if listsAsDuplicateKeys then
           coercedTo singleIniAtom lib.singleton (listOf singleIniAtom) // {
-            description = singleIniAtom.description + " or a list of them for duplicate keys";
+            description = singleIniAtom.description + " or a list of them";
```
Comment (dotlambda) [inline pkgs/pkgs-lib/formats.nix:84]:
```suggestion
            description = singleIniAtom.description + " or a list of them for duplicate keys";
```

**Comment by zhaofengli** [comment]
When values are passed through the module system, somehow everything becomes a list. The end result is confusing:

```nix
let
  pkgs = import ./. {};
  lib = pkgs.lib;

  format = pkgs.formats.ini {
    listToValue = l: lib.concatMapStrings (s: "\n  [list element] ${lib.generators.mkValueStringDefault {} s}") l;
  };

  original = {
    foo = {
      a = "I'm a string";
      b = [
        "and I'm a list"
      ];
    };
  };

  eval = lib.evalModules {
    modules = [
      {
        options = {
          ini = lib.mkOption {
            type = format.type;
          };
        };
        config = {
          ini = original;
        };
      }
    ];
  };
in {
  correct = format.generate "correct.ini" original;
  incorrect = format.generate "incorrect.ini" eval.config.ini;
}
```

`correct.ini`:
```ini
[foo]
a=I'm a string
b=
  [list element] and I'm a list
```

`incorrect.ini`:
```
[foo]
a=
  [list element] I'm a string
b=
  [list element] and I'm a list

```

**Comment by Profpatsch** [comment]
Hm, why was this only added to the `format` instead of the generator?

**Pull request merged**